pkg/apis/gcp/database/v1alpha1/cloudsql_instance_types.go

```diff
@@ -164,33 +176,97 @@ func NewCloudSQLInstanceSpec(properties map[string]string) *CloudsqlInstanceSpec
 			ReclaimPolicy: corev1alpha1.ReclaimRetain,
 		},
 	}
+	spec.AuthorizedNetworks = util.Split(properties["authorizedNetworks"], ",")
+	spec.DatabaseVersion = properties["databaseVersion"]
+	spec.Labels = util.ParseMap(properties["labels"])
+	spec.Region = properties["region"]
+	spec.StorageType = properties["storageType"]
+	if v, err := strconv.Atoi(properties["storageGB"]); err != nil {
+		spec.StorageGB = DefaultStorageGB
+	} else {
+		spec.StorageGB = int64(v)
+	}
+	spec.Tier = properties["tier"]
+	return spec
+}
 
-	val, ok := properties["tier"]
-	if ok {
-		spec.Tier = val
+// ConnectionSecret returns a connection secret for this instance
+func (i *CloudsqlInstance) ConnectionSecret() *corev1.Secret {
+	s := resource.ConnectionSecretFor(i, CloudsqlInstanceGroupVersionKind)
+	s.Data[corev1alpha1.ResourceCredentialsSecretEndpointKey] = []byte(i.Status.Endpoint)
+	s.Data[corev1alpha1.ResourceCredentialsSecretUserKey] = []byte(i.DatabaseUserName())
+	return s
+}
+
+// DatabaseInstance representing spec of this instance
+func (i *CloudsqlInstance) DatabaseInstance(name string) *sqladmin.DatabaseInstance {
```

Comment:
I prefer the separation of having one package defining Crossplane types (with business-logic-free getters and setters as required to satisfy interfaces), and another package (i.e. `clients`) to transform Crossplane types into cloud provider types over implementing this business logic as methods on our core types.

Reply:
I see your point. I tend to lean on another end. I think this logic is closely related to the type, which will produce a `sqladmin.DatabaseInstance`. Different types may produce different versions of `sqladmin.DatabaseInstance`. For example: `CloudsqlInstance` vs. `CloudsqlInstanceFailover` vs. `CloudsqlInstanceReplica`, etc.

Reply:
Yeah, I can't say I think one way is concretely and objectively better than the other.